_posts/2020-06-19-async-explorer.markdown

```diff
@@ -0,0 +1,99 @@
+
+I hope I'll also be able to give you a rough idea of what coroutines are usful for, but it's not the main goal of this post.
+
+This is not an "asyncio" tutorial or anything like that, just me spewing nonsense and messing with some programming languages.
+
+## What is async?
+
+If you're like me, you probably heard terms like "async" and "coroutines" mentioned a lot, and maybe even saw (or wrote) some asynchronious code, but didn't really understand what's going on behind the scenes. Atleast I didn't (and still don't).
+
+Let's consult [Wikipedia](https://en.wikipedia.org/wiki/Coroutine)!
+
+> Coroutines are computer program components that generalize subroutines for non-preemptive multitasking, by allowing execution to be suspended and resumed. 
+
+Let's break it down a little.
+
+For me, a subroutine is a piece of code that accepts parameters, and "executes". That is, processes that parameters, produces side effects, and returns a value. Now, according to Wikipedia, a coroutine is a subroutine that can be "suspended and resumed". The way I understand that, I'm getting a picture of a subroutine "taking a nap", which we can later wake up and continue. Notice that I'm explicitly choosing to not mention multithreading or scheduling. These concepts, which can be similar (and maybe I'll explore that in a future post), are not in my scope currently.
+
+Let's look at an example written in Zig, which shows pretty clearly what we're dealing with:
+
+```zig
+const warn = @import("std").debug.warn;
+
+fn foo(x: i32, y: i32) i32 {
+    warn("1) Hello there! Im taking a nap.\n", .{});
+    suspend;
+
+    warn("3) Why did you wake me up? Im going back to sleep.\n", .{});
+    suspend;
+
+    warn("5) Fine, here you go.\n", .{});
+    return x+y;
+}
+
+fn amain() void {
+    var frame = async foo(1, 2);
+    
+    warn("2) Wake up!\n", .{});
+    resume frame;
+
+    warn("4) Grab a brush and put a little (makeup)!\n", .{});
+    resume frame;
+
+    var ret = await frame;
+    warn("6) foo(1,2) == {}\n", .{ret});
+}
+```
+
+Let's review the flow of this program. We call `foo()` using the `async` keyword, and at that point, `foo()` begins executing. Afterwards, using the keyword `suspend`, `foo()` forfeits it's context, which gives `amain()` an opaque `frame`, which it can use to "wake up" `foo()`. As we can see `foo()` forfeits it's context again, and finally, it returns a return value, using the `return` keyword. At that point, `amain()` retrieves that value using the `await` keyword.
+
+So the expected output is something like this:
+
+```text
+D:\Projects\AsyncExplorer\examples> .\simple_async.exe
+1) Hello there! Im taking a nap.
+2) Wake up!
+3) Why did you wake me up? Im going back to sleep.
+4) Grab a brush and put a little (makeup)!
+5) Fine, here you go.
+6) foo(1,2) == 3
+```
+
+Phew. That was exhausting. But I think we have a pretty solid grip on what the basics of "async" means.
+
+## Examples
```

Comment:
Ur using the same heading level twice in a row... also didn't we just get out of an example?

Reply:
It's WIP, I'll add more examples from different programming languages to this post

Reply:
I've added examples, missing one more Zig example